### PR TITLE
[Impeller] Move Gradient tests to display list.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Migrate almost all of the gradient unittests to display list. Removes tests for disabled dithering and the dithering flag, as it has not been possible to disable dithering on gradients for months and months.

Part of https://github.com/flutter/flutter/issues/142054

### CODE REVIEW

File: impeller/display_list/aiks_dl_gradient_unittests.cc [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+    const DlTileMode tile_modes[] = {DlTileMode::kClamp, DlTileMode::kRepeat,
+                                     DlTileMode::kMirror, DlTileMode::kDecal};
+
+    static int selected_tile_mode = 0;
+    static Matrix matrix = {
+        1, 0, 0, 0,  //
+        0, 1, 0, 0,  //
+        0, 0, 1, 0,  //
+        0, 0, 0, 1   //
+    };
```
Comment (gaaclarke) [inline impeller/display_list/aiks_dl_gradient_unittests.cc:430]:
```suggestion
    static Matrix matrix;
```

File: impeller/golden_tests/golden_playground_test_mac.cc [hunk 2]
```diff
@@ -210,6 +210,26 @@ bool GoldenPlaygroundTest::OpenPlaygroundHere(Picture picture) {
   return SaveScreenshot(std::move(screenshot));
 }
 
+bool GoldenPlaygroundTest::OpenPlaygroundHere(
+    AiksDlPlaygroundCallback
+        callback) {  // NOLINT(performance-unnecessary-value-param)
```
Comment (gaaclarke) [inline impeller/golden_tests/golden_playground_test_mac.cc:215]: Why NOLINT?  Just make this a const reference.

File: impeller/golden_tests/golden_playground_test_stub.cc [hunk 3]
```diff
@@ -37,6 +37,12 @@ bool GoldenPlaygroundTest::OpenPlaygroundHere(
   return false;
 }
 
+bool GoldenPlaygroundTest::OpenPlaygroundHere(
+    AiksDlPlaygroundCallback
+        callback) {  // NOLINT(performance-unnecessary-value-param)
```
Comment (gaaclarke) [inline impeller/golden_tests/golden_playground_test_stub.cc:42]: Same here.

File: impeller/display_list/aiks_dl_gradient_unittests.cc [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+
+  auto gradient = DlColorSource::MakeLinear(
+      {0, 0}, {200, 200}, 2, colors.data(), stops.data(), tile_mode);
+  paint.setColorSource(gradient);
+  paint.setColor(DlColor::kWhite());
+  builder.DrawRect(SkRect::MakeXYWH(0, 0, 600, 600), paint);
+  ASSERT_TRUE(aiks_test->OpenPlaygroundHere(builder.Build()));
+}
+
+static Matrix ToMatrix(const SkMatrix& m) {
```
Comment (gaaclarke) [inline impeller/display_list/aiks_dl_gradient_unittests.cc:52]:
```suggestion
Matrix ToMatrix(const SkMatrix& m) {
```

File: impeller/display_list/aiks_dl_gradient_unittests.cc [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+////////////////////////////////////////////////////////////////////////////////
+// This is for tests of Canvas that are interested the results of rendering
+// gradients.
+////////////////////////////////////////////////////////////////////////////////
+
+namespace impeller {
+namespace testing {
+
+namespace {
+void CanRenderLinearGradient(AiksTest* aiks_test, DlTileMode tile_mode) {
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply impeller/display_list/aiks_dl_gradient_unittests.cc:32]: Or maybe just move it next to the tests.

**Comment by gaaclarke** [comment]
> Are the "AiksTests" that test with DL so that the names are consistent for goldens? Presumably, they can be renamed later.

Yep.

File: impeller/golden_tests/golden_playground_test_stub.cc (hunk 3 above)
Comment (jonahwilliams) [reply impeller/golden_tests/golden_playground_test_stub.cc]: Done

File: impeller/display_list/aiks_dl_gradient_unittests.cc (hunk 4 above)
Comment (jonahwilliams) [reply impeller/display_list/aiks_dl_gradient_unittests.cc]: Done

File: impeller/display_list/aiks_dl_gradient_unittests.cc (hunk 1 above)
Comment (jonahwilliams) [reply impeller/display_list/aiks_dl_gradient_unittests.cc:423]: Done

File: impeller/display_list/aiks_dl_gradient_unittests.cc [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+    const DlTileMode tile_modes[] = {DlTileMode::kClamp, DlTileMode::kRepeat,
+                                     DlTileMode::kMirror, DlTileMode::kDecal};
+
+    static int selected_tile_mode = 0;
+    static Matrix matrix = {
+        1, 0, 0, 0,  //
+        0, 1, 0, 0,  //
+        0, 0, 1, 0,  //
+        0, 0, 0, 1   //
+    };
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/display_list/aiks_dl_gradient_unittests.cc]: Done

File: impeller/display_list/aiks_dl_gradient_unittests.cc [hunk 7]
```diff
@@ -0,0 +1,839 @@
...
+
+  paint.setColorSource(DlColorSource::MakeLinear(
+      {0, 0}, {800, 500}, 2, colors.data(), stops.data(), DlTileMode::kClamp));
+  builder.DrawRect(SkRect::MakeXYWH(0, 0, 800, 500), paint);
+  ASSERT_TRUE(aiks_test->OpenPlaygroundHere(builder.Build()));
+}
+
+TEST_P(AiksTest, CanRenderLinearGradientWithDitheringEnabled) {
+  CanRenderLinearGradientWithDithering(this);
+}  // namespace
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/display_list/aiks_dl_gradient_unittests.cc]: Done

File: impeller/display_list/aiks_dl_gradient_unittests.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/display_list/aiks_dl_gradient_unittests.cc:35]: Done

File: impeller/display_list/aiks_dl_gradient_unittests.cc [hunk 8]
```diff
@@ -0,0 +1,839 @@
...
+namespace {
+void CanRenderLinearGradientManyColors(AiksTest* aiks_test,
+                                       DlTileMode tile_mode) {
+  DisplayListBuilder builder;
+  builder.Scale(aiks_test->GetContentScale().x, aiks_test->GetContentScale().y);
+  DlPaint paint;
+  builder.Translate(100, 100);
+
+  std::vector<DlColor> colors = {
+      DlColor(Color{0x1f / 255.0, 0.0, 0x5c / 255.0, 1.0}.ToARGB()),
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/display_list/aiks_dl_gradient_unittests.cc:228]: as a follow up I want to add a new constructor to DlColor so we can just replace the impeller colors here.

**Review by gaaclarke — APPROVED** [review]
lgtm!

**Pull request merged**